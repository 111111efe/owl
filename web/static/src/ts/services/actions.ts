import CRM from "../widgets/CRM";
import Discuss from "../widgets/Discuss";
import Widget from "../core/Widget";

export interface Action {
  id: number;
  title: string;
  Widget: typeof Widget;
  default?: boolean;
}

const actions: Action[] = [
  { id: 1, title: "Discuss", Widget: Discuss, default: true },
  { id: 2, title: "CRM", Widget: CRM }
];

export default actions;
