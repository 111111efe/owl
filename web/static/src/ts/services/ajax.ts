export type RPC = (route: string, params: any) => Promise<any>;

export const rpc: RPC = async function(route, params) {
  return (<any>window).demoData.mockAjax(route, params);
};
