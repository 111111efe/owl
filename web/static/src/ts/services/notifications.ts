import { EventBus as Bus } from "../core/event_bus";

//------------------------------------------------------------------------------
// Types
//------------------------------------------------------------------------------

export interface INotification {
  id: number;
  title: string;
  message: string;
  type: "notification" | "warning";
  sticky: boolean;
}

export type NotificationEvent = "notification_added" | "notification_removed";

export type Callback = (notif: INotification) => void;

export interface INotificationManager {
  add(notif: Partial<INotification>): number;
  close(id: number): void;
  on(event: NotificationEvent, owner: any, callback: Callback): void;
}

//------------------------------------------------------------------------------
// Notification Manager
//------------------------------------------------------------------------------
export class NotificationManager extends Bus implements INotificationManager {
  nextID = 1;
  notifications: { [key: number]: INotification } = {};

  add(notif: Partial<INotification>): number {
    const id = this.nextID++;
    const defaultVals = {
      title: "",
      message: "",
      type: "notification",
      sticky: false
    };
    const notification = Object.assign(defaultVals, notif, { id });
    this.notifications[id] = notification;
    this.trigger("notification_added", notification);
    if (!notification.sticky) {
      setTimeout(() => this.close(id), 2500);
    }
    return id;
  }
  close(id: number) {
    let notification = this.notifications[id];
    if (notification) {
      delete this.notifications[id];
      this.trigger("notification_removed", notification);
    }
  }
}
