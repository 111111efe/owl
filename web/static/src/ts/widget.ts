import { Component, PureComponent } from "./core/component";
import { Env } from "./env";

//------------------------------------------------------------------------------
// Widget classes
//------------------------------------------------------------------------------

export class Widget<Props, State> extends Component<Env, Props, State> {}

export class PureWidget<P, S> extends PureComponent<Env, P, S> {
  constructor(parent, props) {
    super(parent, props);
    (<any>this.__widget__).isMobile = this.env.isMobile;
  }
  shouldUpdate(nextProps: P): boolean {
    if ((<any>this.__widget__).isMobile !== this.env.isMobile) {
      (<any>this.__widget__).isMobile = this.env.isMobile;
      return true;
    }
    return super.shouldUpdate(nextProps);
  }
}
