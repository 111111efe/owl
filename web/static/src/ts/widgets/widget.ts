import { BaseWidget } from "../core/base_widget";
import { Env } from "../env";

export class Widget<Props, State> extends BaseWidget<Env, Props, State> {}

// TODO: move this to PureComponent in core
export class PureWidget<Props, State> extends Widget<Props, State> {
  shouldUpdate(nextProps: Props): boolean {
    for (let k in nextProps) {
      if (nextProps[k] !== this.props[k]) {
        return true;
      }
    }
    return false;
  }
  async updateState(nextState: Partial<State>) {
    for (let k in nextState) {
      if (nextState[k] !== this.state[k]) {
        return;
      }
    }
    return super.updateState(nextState);
  }
}
