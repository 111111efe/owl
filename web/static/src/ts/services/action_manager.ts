import { EventBus } from "../core/event_bus";
import { Widget } from "../core/widget";
import { Env } from "../env";
import { Registry } from "../registry";
import { Type } from "../types";
import { IRouter, Query } from "./router";

//------------------------------------------------------------------------------
// Types
//------------------------------------------------------------------------------

export type ActionRequest = string | number;

export type Context = { [key: string]: any };

export interface CommonActionInfo {
  id: number;
  context: Context;
  title: string;
  target: "current" | "new";
}

export interface ClientActionInfo extends CommonActionInfo {
  type: "client";
  name: string;
  Widget: Type<Widget<Env, {}>>;
}

export interface ActWindowInfo extends CommonActionInfo {
  type: "act_window";
  view: string;
}

export type ActionInfo = ClientActionInfo | ActWindowInfo;
export type ActionStack = ActionInfo[];

export type ActionEvent = "action_stack_updated";

type Callback = (stack: ActionStack) => void;

export interface IActionManager {
  activate(): void;
  doAction(request: ActionRequest): void;
  on(event: ActionEvent, owner: any, callback: Callback): void;
  getStack(): ActionStack;
}

//------------------------------------------------------------------------------
// Action Manager
//------------------------------------------------------------------------------

export class ActionManager extends EventBus implements IActionManager {
  router: IRouter;
  registry: Registry;
  stack: ActionStack;

  constructor(router: IRouter, registry: Registry) {
    super();
    this.router = router;
    this.registry = registry;
    this.stack = [];
  }

  activate() {
    this.router.on("query_changed", this, this.update);
    this.update(this.router.getQuery());
  }
  doAction(request: ActionRequest) {
    if (typeof request === "number") {
      // this is an action ID
      let name = request === 1 ? "discuss" : "crm";
      let title = request === 1 ? "Discuss" : "CRM";
      let Widget = this.registry.getAction(name);
      this.stack = [
        {
          id: 1,
          context: {},
          target: "current",
          type: "client",
          name,
          title,
          Widget: Widget
        }
      ];
      this.trigger("action_stack_updated", this.stack);
    }
  }

  getStack(): ActionStack {
    return [];
  }

  update(query: Query) {
    if ("action_id" in query) {
      const actionID = parseInt(query.action_id);
      this.doAction(actionID);
    }
  }
}
