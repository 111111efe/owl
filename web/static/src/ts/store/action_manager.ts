import { IAjax } from "./ajax";
import { Type } from "../core/component";
import { EventBus } from "../core/event_bus";
import { Registry } from "../core/registry";
import { Widget } from "../widgets/widget";

//------------------------------------------------------------------------------
// Types
//------------------------------------------------------------------------------

export type ActionRequest = string | number;

export type Context = { [key: string]: any };

export interface CommonActionInfo {
  id: number;
  context: Context;
  title: string;
  target: "current" | "new";
}

export type ActionWidget = Type<Widget<{}, {}>>;

export interface ClientActionInfo extends CommonActionInfo {
  type: "client";
  name: string;
  Widget: ActionWidget;
}

export interface ActWindowInfo extends CommonActionInfo {
  type: "act_window";
  view: string;
}

export interface ActionDescription {
  id: number;
  type: "ir.actions.act_window" | "ir.actions.client";
  target: "current";
}

export type ActionInfo = ClientActionInfo | ActWindowInfo;
export type ActionStack = ActionInfo[];

export type ActionEvent = "action_stack_updated";

type Callback = (stack: ActionStack) => void;

export interface IActionManager {
  doAction(request: ActionRequest): void;
  on(event: ActionEvent, owner: any, callback: Callback): void;
  getStack(): ActionStack;
}

//------------------------------------------------------------------------------
// Action Manager
//------------------------------------------------------------------------------

export class ActionManager extends EventBus implements IActionManager {
  registry: Registry<ActionWidget>;
  ajax: IAjax;
  stack: ActionStack;

  constructor(registry: Registry<ActionWidget>, ajax: IAjax) {
    super();
    this.registry = registry;
    this.ajax = ajax;
    this.stack = [];
  }

  doAction(request: ActionRequest) {
    if (typeof request === "number") {
      this.loadAction(request);
      // this is an action ID
      let name = request === 131 ? "discuss" : "crm";
      let title =
        request === 131 ? "Discuss" : request === 250 ? "Notes" : "CRM";
      let Widget = this.registry.get(name);
      this.stack = [
        {
          id: 1,
          context: {},
          target: "current",
          type: "client",
          name,
          title,
          Widget: Widget
        }
      ];
      this.trigger("action_stack_updated", this.stack);
    }
  }

  private loadAction(id: number) {
    this.ajax.rpc({
      route: "web/action/load",
      params: {
        action_id: id
      }
    });
  }

  getStack(): ActionStack {
    return [];
  }
}
