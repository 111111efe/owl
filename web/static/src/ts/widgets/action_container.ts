import { ActionStack } from "../store/store";
import { Widget } from "./widget";

//------------------------------------------------------------------------------
// Types
//------------------------------------------------------------------------------

export interface Props {
  stack: ActionStack;
}

//------------------------------------------------------------------------------
// Action Container
//------------------------------------------------------------------------------

export class ActionContainer extends Widget<Props, {}> {
  template = "web.action_container";
  currentWidget: any;

  willStart() {
    return this.setContentWidget();
  }

  mounted() {
    if (this.currentWidget && this.currentWidget.el) {
      this.el!.appendChild(this.currentWidget.el);
      this.currentWidget.__mount();
    }
  }

  shouldUpdate(nextProps: Props) {
    if (nextProps.stack !== this.props.stack) {
      this.props = nextProps;
      this.setContentWidget();
    }
    return false;
  }

  async setContentWidget() {
    const info = this.props.stack[this.props.stack.length - 1];
    if (info && info.type === "client") {
      const Widget = info.Widget;
      let widget = new Widget(this, {});
      await widget.mount(this.el || document.createElement("div"));
      if (this.currentWidget) {
        this.currentWidget.destroy();
      }
      this.currentWidget = widget;
    }
  }
}
