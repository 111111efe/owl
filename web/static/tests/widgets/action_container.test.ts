import { ActionStack } from "../../src/ts/services/action_manager";
import { ActionContainer, Props } from "../../src/ts/widgets/action_container";
import { Widget } from "../../src/ts/widgets/widget";
import * as helpers from "../helpers";

//------------------------------------------------------------------------------
// Setup and helpers
//------------------------------------------------------------------------------

let fixture: HTMLElement;
let env: ReturnType<typeof helpers.makeTestEnv>;
let props: Props;
let templates: string;

beforeAll(async () => {
  templates = await helpers.loadTemplates();
});

beforeEach(() => {
  fixture = helpers.makeTestFixture();
  env = helpers.makeTestEnv();
  env.qweb.loadTemplates(templates);
  props = { stack: [] };
});

afterEach(() => {
  fixture.remove();
});

class ClientAction extends Widget<{}, {}> {
  inlineTemplate = "<div>some client action</div>";
}

const demoStack: ActionStack = [
  {
    id: 33,
    context: {},
    title: "some title",
    target: "new",
    type: "client",
    name: "hey",
    Widget: ClientAction
  }
];

//------------------------------------------------------------------------------
// Tests
//------------------------------------------------------------------------------

test("can be rendered with an empty stack", async () => {
  const container = new ActionContainer(env, props);
  await container.mount(fixture);
  expect(fixture.innerHTML).toMatchSnapshot();
});

test("can be rendered with a non empty stack", async () => {
  props.stack = demoStack;
  const container = new ActionContainer(env, props);
  await container.mount(fixture);
  expect(fixture.innerHTML).toMatchSnapshot();
});

test("content is updated properly when new props are given", async () => {
  const container = new ActionContainer(env, props);
  await container.mount(fixture);
  await container.updateProps({ stack: demoStack });
  await helpers.nextTick();
  expect(fixture.innerHTML).toMatchSnapshot();
});
