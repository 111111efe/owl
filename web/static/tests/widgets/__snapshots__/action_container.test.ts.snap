// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can be rendered with a non empty stack 1`] = `"<div class=\\"o_content\\"><div>some client action</div></div>"`;

exports[`can be rendered with an empty stack 1`] = `"<div class=\\"o_content\\"></div>"`;

exports[`content is updated properly when new props are given 1`] = `"<div class=\\"o_content\\"><div>some client action</div></div>"`;
